tests/unit_tests/udfs/test_token_classification.py

```diff
@@ -137,29 +152,59 @@ def create_mock_metadata(udf_wrapper):
 
-    executor = UDFMockExecutor()
-    meta = create_mock_metadata_with_span(params.work_with_span_udf_wrapper)
+    mock_meta = create_mock_metadata_with_span(params.work_with_span_udf_wrapper)
+    input = params.work_with_span_input_data
+    mock_ctx = StandaloneMockContext(inp=input, metadata=mock_meta)
 
-    exa = MockExaEnvironment(
-        metadata=meta,
+
+    mock_exa = MockExaEnvironment(
+        metadata=mock_meta,
         connections=params.bfs_connections)
 
-    result = executor.run([Group(params.work_with_span_input_data)], exa)
-    result_output = Output(result[0].rows)
-    expected_output = Output(params.work_with_span_output_data)
-    n_input_columns = len(meta.input_columns) - 1
+    #--------
+    # do we want to still use this?
+    #executor = UDFMockExecutor()
+    #result = executor.run(mock_ctx, mock_exa) # todo this would need  changes in UDFMockExecutor
+    #--------
+    # or do we want a version like below?
+    # or is there a secret magic power the STandAloneUDFMock has that i am not seeing?
+    mock_base_model_factory: Union[ModelFactoryProtocol, MagicMock] = create_autospec(ModelFactoryProtocol, _name="mock_base_model_factory")
+    mock_models: List[Union[transformers.AutoModel, MagicMock]] = [
+        create_autospec(transformers.AutoModel)
+        ]
+    mock_cast(mock_base_model_factory.from_pretrained).side_effect = mock_models
+
+    mock_tokenizer_factory: Union[ModelFactoryProtocol, MagicMock] = create_autospec(ModelFactoryProtocol)
+    mock_pipeline:  List[Union[transformers.AutoModel, MagicMock]] = [
+        create_autospec(Pipeline, side_effect=[params.tokenizer_model_output_df])
+        ]
+    mock_pipeline_factory: Union[Pipeline, MagicMock] = create_autospec(Pipeline, side_effect=mock_pipeline) #todo do we want to use params.mock_pipeline instead?
+    # todo do we want to use the udf or the mock in params?
+    udf = TokenClassificationUDF(exa=mock_exa,
+                                 base_model=mock_base_model_factory,
+                                 tokenizer=mock_tokenizer_factory,
+                                 pipeline=mock_pipeline_factory,
+                                 work_with_spans=True)
+
+    udf.run(mock_ctx)
+    result = mock_ctx.output
+
+    assert result[0][-1] is None and len(result[0]) == len(mock_meta.output_columns)
+
+    expected_output = mock_meta.output_columns
+    n_input_columns = len(mock_meta.input_columns) - 1
     try:
         assert (
-            OutputMatcher(result_output, n_input_columns) == expected_output,
-            params.mock_pipeline.counter == params.expected_model_counter)
+            OutputMatcher(result, n_input_columns) == expected_output, #validate emit checks output type but not content, but output matcher does not work with the result from mock_ctx
+            params.mock_pipeline.counter == params.expected_model_counter) # pipeline counter not incresed in mock
```

Comment:
you can now use a normal unittest.Mock, where you can get all calls with mock.mock_calls